fvm/src/syscalls/bind.rs

```diff
@@ -95,34 +96,35 @@ fn memory_and_data<'a, K: Kernel>(
     Ok((Memory::new(mem), data))
 }
 
-fn charge_exec_units_for_gas(caller: &mut Caller<InvocationData<impl Kernel>>) -> Result<(), Trap> {
-    let exec_units = caller
-        .data_mut()
-        .calculate_exec_units_for_gas()
-        .map_err(|_| Trap::new("failed to calculate exec_units"))?;
-    if exec_units.is_negative() {
-        caller.add_fuel(u64::try_from(exec_units.saturating_neg()).unwrap_or(0))?;
-    } else {
-        caller.consume_fuel(u64::try_from(exec_units).unwrap_or(0))?;
-    }
+fn gastracker_to_wasmgas(caller: &mut Caller<InvocationData<impl Kernel>>) -> Result<(), Trap> {
+    let avail_gas = caller.data_mut().kernel.get_gas();
+    let frgas = gas::gas_to_frgas(avail_gas);
 
-    let gas_used = caller.data().kernel.gas_used();
-    let fuel_consumed = caller
-        .fuel_consumed()
-        .ok_or_else(|| Trap::new("expected to find exec_units consumed"))?;
-    caller.data_mut().set_snapshots(gas_used, fuel_consumed);
-    Ok(())
+    let gas_global = caller.data_mut().avail_gas_global.unwrap();
+
+    gas_global
+        .set(caller, Val::I64(frgas))
+        .map_err(|_| Trap::new("failed to set available gas"))
 }
 
-fn charge_gas_for_exec_units(caller: &mut Caller<InvocationData<impl Kernel>>) -> Result<(), Trap> {
-    let exec_units_consumed = caller
-        .fuel_consumed()
-        .ok_or_else(|| Trap::new("expected to find exec_units consumed"))?;
+fn wasmgas_to_gastracker(caller: &mut Caller<InvocationData<impl Kernel>>) -> Result<(), Trap> {
+    let global = caller.data_mut().avail_gas_global;
+
+    let frgas = match global.unwrap().get(&mut *caller) {
+        Val::I64(g) => Ok(g),
+        _ => Err(Trap::new("failed to get wasm gas")),
+    }?;
+
+    let available_gas = gas::frgas_to_gas(frgas, false);
 
+    // todo do we have consts for charge names anywhere?
+    // todo make sure that we handle traps/out-of-gas here correctly
```

Comment:
I'm 60% convinced that we're not doing this correctly

Reply:
We're not. If we're out of gas, we need to return an OutOfGas error. We can do that by converting an OutOfGas `Abort` into a `Trap`.